debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+			s.d.sendEvent(Event{Type: StoppedEventType, Thread: t.id, Message: "step", stackIndex: i, stackEvent: e})
+		}
+	}
+	if t.done() {
+		s.d.sendEvent(Event{Type: ThreadEventType, Thread: t.id, Message: "exited"})
+		allStopped := true
+		for _, t := range s.threads {
+			if !t.done() {
+				allStopped = false
+				break
+			}
+		}
+		if allStopped {
+			s.d.sendEvent(Event{Type: TerminatedEventType})
+		}
+	}
+
+	return err
+}
+
+func (s *session) StepOut(threadID ThreadID) error {
+	if s == nil {
+		return fmt.Errorf("no active debug session")
+	}
+
+	t, err := s.thread(threadID)
+	if err != nil {
+		return err
+	}
+
+	err = t.stepOut()
+	if err == nil {
+		i, e, _ := t.current()
+		if e != nil {
+			s.d.sendEvent(Event{Type: StoppedEventType, Thread: t.id, Message: "step", stackIndex: i, stackEvent: e})
+		}
+	}
+	if t.done() {
+		s.d.sendEvent(Event{Type: ThreadEventType, Thread: t.id, Message: "exited"})
+		allStopped := true
+		for _, t := range s.threads {
+			if !t.done() {
+				allStopped = false
+				break
+			}
+		}
+		if allStopped {
+			s.d.sendEvent(Event{Type: TerminatedEventType})
+		}
+	}
+
+	return err
+}
+
+func (s *session) Threads() ([]Thread, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	threads := make([]Thread, 0, len(s.threads))
+	for _, t := range s.threads {
+		threads = append(threads, t)
+	}
+
+	return threads, nil
+}
+
+type sessionThreadState struct {
+	entered     bool
+	ended       bool
+	prevQueryID uint64
+}
+
+func (s *sessionThreadState) String() string {
+	return fmt.Sprintf("{entered: %v, ended: %v, prevQueryID: %d}", s.entered, s.ended, s.prevQueryID)
+}
+
+func (s *session) handleEvent(t *thread, stackIndex int, e *topdown.Event, ts threadState) (eventAction, threadState, error) {
+	state, ok := ts.(*sessionThreadState)
+	if state != nil && !ok {
+		s.d.logger.Warn("invalid thread state: %v", s)
+	}
+	if state == nil {
+		state = &sessionThreadState{}
+	}
+
+	defer func() {
+		if e != nil {
+			state.prevQueryID = e.QueryID
+		} else {
+			state.prevQueryID = 0
+		}
+	}()
+
+	if e == nil {
+		handleEnd := func() (eventAction, threadState, error) {
+			_ = t.close()
+			return stopAction, state, nil
+		}
+
+		if state.ended {
+			s.d.logger.Debug("End of trace already handled")
+			return handleEnd()
+		}
+
+		s.d.logger.Debug("Handling end of trace")
+
+		state.ended = true
+		if s.properties.StopOnResult {
+			s.d.logger.Info("Thread %d stopped at end of trace", t.id)
+			s.d.sendEvent(Event{Type: StoppedEventType, Thread: t.id, Message: "result", stackIndex: stackIndex, stackEvent: e})
+			return breakAction, state, nil
+		}
+
+		return handleEnd()
+	}
+
+	if e.Location == nil {
+		s.d.logger.Debug("Handling event:\n%v\n\nstate:\n%s", e, state)
+	} else {
+		s.d.logger.Debug("Handling event:\n%v\n\nloc:\n%s\n\nstate:\n%s", e, e.Location, state)
+	}
+
+	if s.skipOp(e.Op) {
+		// FIXME: Should we only skip an event as long as we're within the same query scope?
```

Comment:
What's the context here? Would be good to fix this.

Reply:
This was an old comment we can remove. I think we're still gonna need to skip "undesired" operations if they're the next op when entering or leaving a new query level/context.